atom/renderer/api/atom_api_renderer_ipc.cc

```diff
@@ -27,40 +33,88 @@ RenderFrame* GetCurrentRenderFrame() {
   return RenderFrame::FromWebFrame(frame);
 }
 
-void Send(mate::Arguments* args,
-          bool internal,
-          const std::string& channel,
-          const base::ListValue& arguments) {
-  RenderFrame* render_frame = GetCurrentRenderFrame();
-  if (render_frame == nullptr)
-    return;
-
-  bool success = render_frame->Send(new AtomFrameHostMsg_Message(
-      render_frame->GetRoutingID(), internal, channel, arguments));
-
-  if (!success)
-    args->ThrowError("Unable to send AtomFrameHostMsg_Message");
-}
-
-base::ListValue SendSync(mate::Arguments* args,
-                         bool internal,
-                         const std::string& channel,
-                         const base::ListValue& arguments) {
-  base::ListValue result;
-
-  RenderFrame* render_frame = GetCurrentRenderFrame();
-  if (render_frame == nullptr)
+class IPCRenderer : public mate::Wrappable<IPCRenderer> {
+ public:
+  explicit IPCRenderer(v8::Isolate* isolate) {
+    Init(isolate);
+    RenderFrame* render_frame = GetCurrentRenderFrame();
+    DCHECK(render_frame);
+    render_frame->GetRemoteInterfaces()->GetInterface(
+        mojo::MakeRequest(&electron_browser_ptr_));
+  }
+  static void BuildPrototype(v8::Isolate* isolate,
+                             v8::Local<v8::FunctionTemplate> prototype) {
+    prototype->SetClassName(mate::StringToV8(isolate, "IPCRenderer"));
+    mate::ObjectTemplateBuilder(isolate, prototype->PrototypeTemplate())
+        .SetMethod("send", &IPCRenderer::Send)
+        .SetMethod("sendSync", &IPCRenderer::SendSync);
+  }
+  static mate::Handle<IPCRenderer> Create(v8::Isolate* isolate) {
+    return mate::CreateHandle(isolate, new IPCRenderer(isolate));
+  }
+
+  void Send(mate::Arguments* args,
+            bool internal,
+            const std::string& channel,
+            const base::ListValue& arguments) {
+    electron_browser_ptr_->Message(internal, channel, arguments.Clone());
+  }
+
+  base::Value SendSync(mate::Arguments* args,
+                       bool internal,
+                       const std::string& channel,
+                       const base::ListValue& arguments) {
+    base::Value result;
+
+    // A task is posted to a separate thread to execute the request so that
+    // this thread may block on a waitable event. It is safe to pass raw
+    // pointers to |sync_load_response| and |event| as this stack frame will
+    // survive until the request is complete.
+    scoped_refptr<base::SingleThreadTaskRunner> task_runner =
+        base::CreateSingleThreadTaskRunnerWithTraits({});
+
+    base::WaitableEvent response_received_event;
+
+    auto interface_info = electron_browser_ptr_.PassInterface();
+    atom::mojom::ElectronBrowserPtrInfo returned_interface_info;
+    task_runner->PostTask(
+        FROM_HERE,
+        base::BindOnce(
+            [](atom::mojom::ElectronBrowserPtrInfo interface_info,
```

Comment:
Not sure why capturing lambdas aren't allowed via `base::Bind` (wild guess it might be the **cons** described in https://google.github.io/styleguide/cppguide.html#Lambda_expressions), but they seem to available for tests with help of [base::BindLambadForTesting](https://cs.chromium.org/chromium/src/base/test/bind_test_util.h?l=29-36), using a similar utility can simplify this block. In this case the execution is known to be blocked, so there shouldn't any worry of dangling pointers.

```c++
    task_runner->PostTask(
        FROM_HERE,
        base::BindOnce(
            BindLambda(
                [&](atom::mojom::ElectronBrowserPtrInfo interface_info) {
                  atom::mojom::ElectronBrowserPtr browser_ptr(
                      std::move(interface_info));
                  browser_ptr->MessageSync(
                      internal, channel, arguments.Clone(),
                      base::BindOnce(
                          BindLambda(
                              [&](atom::mojom::ElectronBrowserPtr ptr,
                                  base::Value response) {
                                result = std::move(response);
                                returned_interface_info = ptr.PassInterface();
                                response_received_event.Signal();
                              }),
                          std::move(browser_ptr)));
                }),
            std::move(interface_info)));
```

its just that nested lambdas doesn't look good at first pass :) even binding them as helper functions would be good.

Reply:
I'm hesitant to use a utility that's specifically marked for testing only in a non-test code path. I'll split these out into helper functions to make this a little clearer.